### PR TITLE
[NEW MODEL CLIP] Add disco diffusion clip vitb32

### PR DESCRIPTION
**Pull request by JunnYu**
### PR types

New features
### PR changes

Models
### Description

1. 新增CLIPModel、CLIPTextModel、CLIPVisionModel
2. CLIPModel + DD生成图片。 V100 32G 10分钟一张
3. CLIPModel + StableDiffusion 的3种文图生成模式。V100 32G 7秒一张
  - Text-to-Image Generation
  - Image-to-Image Text-Guided Generation
  - Text-Guided Image Inpainting

```python
import paddle
paddle.set_device("gpu:5")
from paddlenlp.transformers import CLIPForImageGeneration, CLIPTokenizer

# Initialize the model and tokenizer
model_name_or_path = 'openai/disco-diffusion-clip-vit-base-patch32'
model = CLIPForImageGeneration.from_pretrained(model_name_or_path)
tokenizer = CLIPTokenizer.from_pretrained(model_name_or_path)
model.eval()

# Prepare the model inputs.
prompts = "A beautiful painting of a singular lighthouse, shining its light across a tumultuous sea of blood by greg rutkowski and thomas kinkade, Trending on artstation."
tokenizer.pad_token_id = 0
tokenized_inputs = tokenizer(prompts, return_tensors="pd", padding="max_length", max_length=77)

images = model.generate(**tokenized_inputs)
# return List[PIL.Image]
images[0].save("figure.png")
```
下载

### CODE REVIEW

File: paddlenlp/transformers/clip/feature_extraction.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/feature_extraction.py:1]: 这两个Copyright的顺序换一下吧，自己的写在最上面

File: paddlenlp/transformers/clip/modeling.py [hunk 2]
```diff
@@ -0,0 +1,1417 @@
...
+        if position_ids is None:
+            position_ids = self.position_ids[:, :seqlen]
+
+        causal_mask = self.causal_mask[:, :, :seqlen, :seqlen]
+        if attention_mask is not None:
+            if attention_mask.ndim == 2:
+                attention_mask = attention_mask[:, None, None, :]
+            attention_mask = attention_mask + causal_mask
+        else:
+            attention_mask = causal_mask
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py:680]: 新增模型就按照对齐HF的0/1语义来实现mask吧

**Review by rainyfly — DISMISSED** [review]
LGTM for diffusion_generate process.

File: paddlenlp/transformers/clip/modeling.py [hunk 3]
```diff
@@ -0,0 +1,1417 @@
...
+            return (last_hidden_state, pooled_output) + encoder_outputs[1:]
+
+        return BaseModelOutputWithPoolingAndCrossAttentions(
+            last_hidden_state=last_hidden_state,
+            pooler_output=pooled_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions)
+
+
+class TextTransformer(CLIPPreTrainedModel):
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py]: 这里能否跟HF一样也叫 CLIPTextTransformer 呢，vision的那个同样

File: paddlenlp/transformers/clip/modeling.py [hunk 4]
```diff
@@ -0,0 +1,1417 @@
...
+def quick_gelu(x):
+    return x * F.sigmoid(1.702 * x)
+
+
+F.quick_gelu = quick_gelu
+
+INF = float("-inf")  # -1e4 -1e9
+
+
+class VisionTransformer(CLIPPreTrainedModel):
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py]: 这个是否也和HF一样使用 CLIPVisionTransformer

File: paddlenlp/transformers/clip/modeling.py [hunk 5]
```diff
@@ -0,0 +1,1417 @@
...
+
+    def __init__(self,
+                 input_resolution: int,
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py:544]: input_resolution 是否使用 image_size 呢，看HF是用的image_size，哪种更常用更好些呢

File: paddlenlp/transformers/clip/modeling.py [hunk 6]
```diff
@@ -0,0 +1,1417 @@
...
+            save_rate (int, optional):
+                During a diffusion run, you can monitor the progress of each image being created with this variable.  If display_rate is set 
+                to 50, DD will show you the in-progress image every 50 timesteps. Setting this to a lower value, like 5 or 10, is a good way 
+                to get an early peek at where your image is heading. If you don't like the progression, just interrupt execution, change some
+                settings, and re-run.  If you are planning a long, unmonitored batch, it's better to set display_rate equal to steps, because 
+                displaying interim images does slow Colab down slightly.
+                Default to `10`.
+            n_batches (int, optional):
+                This variable sets the number of still images you want DD to create.  If you are using an animation mode (see below for details)
+                DD will ignore n_batches and create a single set of animated frames based on the animation settings.
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py:1778]: animation mode也是可以支持的参数吗

File: paddlenlp/transformers/clip/modeling.py [hunk 7]
```diff
@@ -0,0 +1,1417 @@
...
+            dropout=0.0,
+            resblock_updown=True,
+            use_new_attention_order=False,
+        )
+        self.secondary_model = create_secondary_model()
+
+        # eval mode and stop all param's gradient
+        self.eval()
+        for param in self.parameters():
+            param.stop_gradient = True
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py:1452]: 感觉整个的stop_gradient加到这里会不会有点奇怪，和别的也有点不统一

File: paddlenlp/transformers/clip/modeling.py [hunk 8]
```diff
@@ -0,0 +1,1417 @@
...
+            raw_pad_token_id = tokenizer.pad_token_id
+            tokenizer.pad_token_id = 0
+            tokenized_inputs = tokenizer(text_prompt, return_tensors="pd", padding="max_length", max_length=tokenizer.model_max_length)
+            images = model.generate(**tokenized_inputs)
+            
+            # return List[PIL.Image]
+            images[0].save("figure.png")
+                
+        """
+        self.diffusion = create_gaussian_diffusion(
```
Comment (guoshengCS) [inline paddlenlp/transformers/clip/modeling.py:1826]: 这个create_gaussian_diffusion是否也作为DiffusionMixin的成员方法呢

File: paddlenlp/transformers/guided_diffusion_utils/diffusion_utils.py [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+                init = create_perlin_noise([1.5**-i * 0.5 for i in range(12)],
+                                           1, 1, False, side_y, side_x)
+                init2 = create_perlin_noise([1.5**-i * 0.5 for i in range(8)],
+                                            4, 4, True, side_y, side_x)
+            init = (T.to_tensor(init).add(T.to_tensor(init2)).divide(
+                paddle.to_tensor(2.0)).unsqueeze(0) * 2 - 1)
+            del init2
+
+        if init is not None and init_scale:
+            lpips = try_import("paddle_lpips")
```
Comment (guoshengCS) [inline paddlenlp/transformers/guided_diffusion_utils/diffusion_utils.py]: paddle_lpips这个是什么呢

File: paddlenlp/transformers/clip/modeling.py (hunk 2 above)
Comment (JunnYu) [reply paddlenlp/transformers/clip/modeling.py:680]: done

File: paddlenlp/transformers/clip/feature_extraction.py (hunk 1 above)
Comment (JunnYu) [reply paddlenlp/transformers/clip/feature_extraction.py:1]: done

**Comment by JunnYu** [comment]
### 添加stable diffusion使用

#### 1、文本生成图片 Text-to-Image Generation
```python
import paddle
paddle.set_device("gpu:5")
from PIL import Image
from IPython.display import display
from paddlenlp.transformers import CLIPModel, CLIPTokenizer, CLIPForImageGeneration

model = CLIPForImageGeneration.from_pretrained("CompVis/stable-diffusion-v1-4")
tokenizer = CLIPTokenizer.from_pretrained("CompVis/stable-diffusion-v1-4")

# text2image
prompts = [
    "In the morning light,Chinese ancient buildings in the mountains,Magnificent and fantastic John Howe landscape,lake,clouds,farm,Fairy tale,light effect,Dream,Greg Rutkowski,James Gurney,artstation",
    "clouds surround the mountains and Chinese palaces,sunshine,lake,overlook,overlook,unreal engine,light effect,Dream，Greg Rutkowski,James Gurney,artstation",
    "close-up maximalist illustration of panther, by makoto shinkai, akihiko yoshida, yoshitaka amano, super detailed, hd wallpaper, digital art",
    "in the morning light,Overlooking TOKYO city by greg rutkowski and thomas kinkade,Trending on artstationmakoto shinkai style",
]
all_images = []
for prompt in prompts:
    tokenized_inputs = tokenizer(prompt, padding="max_length", trunction=True, max_length=tokenizer.model_max_length, return_tensors="pd")
    seed = 42
    image = model.generate(tokenized_inputs["input_ids"], mode="text2image", seed=seed)[0]
    display(prompt)
    display(image)
    all_images.append(image)
```
f_0
f_1
f_2
f_3

#### 2、文本+图片继续创作 Image-to-Image Text-Guided Generation
```python
import paddle
paddle.set_device("gpu:5")
from PIL import Image
from IPython.display import display
from paddlenlp.transformers import CLIPModel, CLIPTokenizer, CLIPForImageGeneration

model = CLIPForImageGeneration.from_pretrained("CompVis/stable-diffusion-v1-4")
tokenizer = CLIPTokenizer.from_pretrained("CompVis/stable-diffusion-v1-4")

def download_image(url):
    import requests
    from io import BytesIO
    response = requests.get(url)
    init_img = Image.open(BytesIO(response.content)).convert("RGB")
    return init_img

prompt = "A fantasy landscape, trending on artstation"
tokenized_inputs = tokenizer(prompt, padding="max_length", trunction=True, max_length=tokenizer.model_max_length, return_tensors="pd")
init_image = "http://bj.bcebos.com/paddlenlp/models/community/CompVis/stable-diffusion-v1-4/sketch-mountains-input.png"
init_image = download_image(init_image)
display(init_image)
seed = 42
image = model.generate(tokenized_inputs["input_ids"], mode="image2image", init_image=init_image, seed=seed, strength=0.75, guidance_scale=7.5)[0]
display(image) 
```
init_image
new

#### 3、补全图像的某个部分 Text-Guided Image Inpainting
```python
import paddle
paddle.set_device("gpu:5")
from PIL import Image
from IPython.display import display
from paddlenlp.transformers import CLIPModel, CLIPTokenizer, CLIPForImageGeneration

model = CLIPForImageGeneration.from_pretrained("CompVis/stable-diffusion-v1-4")
tokenizer = CLIPTokenizer.from_pretrained("CompVis/stable-diffusion-v1-4")

seed = 42
prompt = "a cat sitting on a bench"
init_image = "http://bj.bcebos.com/paddlenlp/models/community/CompVis/stable-diffusion-v1-4/overture-creations.png"
mask_image = "http://bj.bcebos.com/paddlenlp/models/community/CompVis/stable-diffusion-v1-4/overture-creations-mask.png"

def download_image(url):
    import requests
    from io import BytesIO
    response = requests.get(url)
    init_img = Image.open(BytesIO(response.content)).convert("RGB")
    return init_img

init_image = download_image(init_image)
mask_image = download_image(mask_image).convert("L")
tokenized_inputs = tokenizer(prompt, padding="max_length", trunction=True, max_length=tokenizer.model_max_length, return_tensors="pd")
image = model.generate(tokenized_inputs["input_ids"], mode="inpaint", init_image=init_image, mask_image=mask_image, seed=seed, strength=0.75)[0]
display("init_image ----------------------->")
display(init_image)
display("mask_image ----------------------->")
display(mask_image)
display("new_image ----------------------->")
display(image)
```
init
mask
new (2)

**Review by guoshengCS — DISMISSED** [review]

**Review by guoshengCS — APPROVED** [review]
